core/src/main/java/org/apache/carbondata/core/dictionary/client/DictionaryClientHandler.java

```diff
@@ -77,20 +79,13 @@ public void exceptionCaught(ChannelHandlerContext ctx, ExceptionEvent e) throws
    */
   public DictionaryKey getDictionary(DictionaryKey key) {
     DictionaryKey dictionaryKey;
-    BlockingQueue<DictionaryKey> dictKeyQueue = null;
     try {
-      synchronized (lock) {
-        dictKeyQueue = dictKeyQueueMap.get(key.getThreadNo());
-        if (dictKeyQueue == null) {
-          dictKeyQueue = new LinkedBlockingQueue<DictionaryKey>();
-          dictKeyQueueMap.put(key.getThreadNo(), dictKeyQueue);
-        }
-      }
-      byte[] serialize = KryoRegister.serialize(key);
-      ctx.getChannel().write(serialize);
+      ByteBuf buffer = ctx.alloc().buffer();
+      key.writeData(buffer);
+      ctx.writeAndFlush(buffer);
     } catch (Exception e) {
-      LOGGER.error("Error while send request to server " + e.getMessage());
-      ctx.getChannel().close();
+      LOGGER.error(e, "Error while send request to server " + e.getMessage());
+      ctx.close();
     }
     boolean interrupted = false;
     try {
```

Comment:
There is no timeout control here. How about using SettingFuture from Guava libary, like
```
final SettableFuture<Integer> dictionaryKey = SettableFuture.create();
    ResponseCallback callback = new ResponseCallback() {
      @Override
      public void invoke(NettyMessage response) {
        if (response instanceof DictGenResponseMessage) {
          DictGenResponseMessage resp = (DictGenResponseMessage) response;
          dictionaryKey.set(resp.getDictionaryKey());
        }
      }
    };
    clientHandler.addRequest(request.id(), callback);
    ChannelFuture f = channel.writeAndFlush(request);
    try {
      return dictionaryKey.get(10, TimeUnit.SECONDS);
    } catch (InterruptedException | TimeoutException | ExecutionException e) {
      LOGGER.error("exception while generating dictionary key from Dictionary Service: " +
          e.getMessage());
      throw e;
    }
```
This needs to set a callback in the handler and invoke the callback in `channelRead0` in the handler. Then the blockingQueue is not needed